### PR TITLE
[MRG] BSP-OT: Sparse transport plans between discrete measures in loglinear time

### PR DESCRIPTION
**Pull request by dcoeurjo**
Initial fork and PR. (wip) 

## Types of changes

[Original paper](https://perso.liris.cnrs.fr/nicolas.bonneel/paper_BSPflow.pdf)

## PR checklist

- [ ] I have read the [**CONTRIBUTING**](https://pythonot.github.io/contributing.html) document.
- [ ] The documentation is up-to-date with the changes I made (check build artifacts).
- [ ] All tests passed, and additional code has been **covered with new tests**.
- [ ] I have added the PR and Issue fix to the [**RELEASES.md**](RELEASES.md) file.

### CODE REVIEW

File: ot/bsp/BSP-OT_header_only.h [hunk 1]
```diff
@@ -11,11 +11,11 @@
 #include <vector>
 #include <Eigen/Sparse>
 
-#include <filesystem>
+// #include <filesystem>
```
Comment (dcoeurjo) [inline ot/bsp/BSP-OT_header_only.h]: clean dead code ?

File: ot/bsp/BSP-OT_header_only.h [hunk 2]
```diff
@@ -0,0 +1,2842 @@
...
+            std::cerr << "end - beg == 0" << std::endl;;
+            //return;
+        }
+        if (gap == 1){
+            return;
+        }
+        int pivot = random_pivot ? randint(beg+1,end-1) : beg + (end-beg)/2;
+        Vector<D> d = getSlice(idA,idB,beg,end);
+        d -= d.dot(prev_slice)*prev_slice/prev_slice.squaredNorm();
+        d.normalized();
```
Comment (nbonneel) [inline ot/bsp/BSP-OT_header_only.h]: ====> d.normalize()  !

**Review by nbonneel — CHANGES_REQUESTED** [review]

File: ot/bsp/BSP-OT_header_only.h (hunk 2 above)
Comment (baptiste-genest) [reply ot/bsp/BSP-OT_header_only.h]: done

File: test/test_bspot.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    cost_ot = np.sum(G * M)
+
+    # check that the relative error is small
+    relative_error = np.abs(cost - cost_ot) / (cost_ot + 1e-8)
+    assert relative_error < 0.1, "Relative error is too large: {}".format(
+        relative_error
+    )
+
+
+def test_bsp_ot_torch_backend():
```
Comment (rflamary) [inline test/test_bspot.py:114]: add pytest skipif if torch not installed

File: test/test_ot_openmp.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import ot
+import numpy as np
+
+
+def test_omp_emd2():
```
Comment (rflamary) [inline test/test_ot_openmp.py]: remove this test it alerady exists

File: test/test_bspot.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    A_new = A - A_torch.grad.detach().numpy() * n / 2
+
+    # compute new cost with advected points
+    cost_new, _, _ = ot.bsp.compute_bspot_bijection(A_new, B, 1)
+
+    # cost should be zero
+    np.testing.assert_allclose(cost_new, 0, atol=1e-5)
+
+
+test_bsp_ot_exact_identity()
```
Comment (rflamary) [inline test/test_bspot.py]: you should not call the tests

File: docs/source/all.rst [hunk 6]
```diff
@@ -39,6 +39,7 @@ API and modules
    unbalanced
    utils
    weak
+   bsp
```
Comment (rflamary) [inline docs/source/all.rst]: move this one up to stay alphabetical please

File: examples/plot_bsp_ot.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
+"""
+===================
+Fast and accurate bijections using BSP-OT
+===================
+
+This example shows two use cases for the bijections provided by BSP-OT,
+between two large point clouds: shape morphing (and animated morphing)
+and full color transfer (pixel permutation).
+
+[?] Genest, B., Bonneel, N., Nivoliers, V., Coeurjolly, D.
```
Comment (rflamary) [inline examples/plot_bsp_ot.py]: the number  ? is the same as in the end of the README file where you should add the reference also ;)

You should also have the same reference in the functions docstring with same number (and reference the algorithm in the paper from teh doc)

File: examples/plot_bsp_ot.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    X2 = sample_ball(n_half, radius, centers[1])
+
+    return np.vstack((X1, X2))
+
+
+# ----------------------------
+# Load point clouds
+# ----------------------------
+N = 100000
+A = sample_ball(N)
```
Comment (rflamary) [inline examples/plot_bsp_ot.py]: could you use X_s and X_t ? A and B are used for sample weights in other exmaples

File: examples/plot_bsp_ot.py [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+B = sample_two_balls(N, 0.5)
+
+N = A.shape[0]
+
+
+##############################################################################
+# Bijection computation
+# ----------------------------------
+
+start = time.time()
```
Comment (rflamary) [inline examples/plot_bsp_ot.py]:
```suggestion
ot.tic()
```

File: examples/plot_bsp_ot.py [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+
+# %% call BSP-OT solver
+# The solver returns the transport cost, the final bijection and the
+# intermediary ones used to compute the final one (here we set k = 64).
+# Here we only use the final bijection.
+cost, perm, _ = ot.bsp.compute_bspot_bijection(A, B, 64, 2)
+print(
+    "Bijection computed between {} points, with cost {} in {}s".format(
+        N, cost, time.time() - start
```
Comment (rflamary) [inline examples/plot_bsp_ot.py]:
```suggestion
        N, ot.toq()
```
return teh tiem wit better precision

File: examples/plot_bsp_ot.py [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+
+
+I1 = np.asarray(I1_pil).astype(np.float64) / 255.0
+I2 = np.asarray(I2_pil).astype(np.float64) / 255.0
+
+X1 = im2mat(I1)
+X2 = im2mat(I2)
+
+
+start = time.time()
```
Comment (rflamary) [inline examples/plot_bsp_ot.py]: same here

File: ot/bsp/bsp_ot.py [hunk 12]
```diff
@@ -0,0 +1,162 @@
+import numpy as np
+import warnings
+
+from ..utils import list_to_array
+from ..backend import get_backend
+from .bsp_wrap import bsp_solve_c, merge_bijections_c
+
+
+def compute_bspot_bijection(
+    X, Y, n_plans=64, lp_power=2, initial_plan=None, gaussian=True
```
Comment (rflamary) [inline ot/bsp/bsp_ot.py]:
```suggestion
    X, Y, n_plans=64, lp_power=2, initial_plan=None, gaussian_slicing='auto'
```

put a clearer name and have an automatic decison wrt d ? and allow for True or False

File: ot/bsp/bsp_ot.py [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+        If true then uses the Gaussian slicing heuristic to improve matching quality.
+        Comes with a cubic complexity with dimension, set at true by default.
+
+    Returns
+    -------
+    cost : float
+        The transport cost of the final bijection.
+    plan : array-like, shape (n_samples,)
+        The final bijection, stored as a permutation (e.g. a list of numbers) such that X[i] is assigned to Y[plan[i]].
+    plans : list of array
```
Comment (rflamary) [inline ot/bsp/bsp_ot.py]: this should be array_like and size should be given (n_samples,n_plans)

File: ot/bsp/bsp_ot.py [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+        Bijection to use for initializing merging (optional).
+    gaussian : bool, optional
+        If true then uses the Gaussian slicing heuristic to improve matching quality.
+        Comes with a cubic complexity with dimension, set at true by default.
+
+    Returns
+    -------
+    cost : float
+        The transport cost of the final bijection.
+    plan : array-like, shape (n_samples,)
```
Comment (rflamary) [inline ot/bsp/bsp_ot.py]: call it perm, or perm_plan so that it is clear that it is a permutation, plan in POT is a OT plan (that  can be sparse)

File: ot/bsp/bsp_ot.py [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+        warnings.warn(
+            "Cost computed from plan differs from cost returned by solver. Cost from plan: {}, Cost from solver: {}".format(
+                cost, cost_inner
+            )
+        )
+
+    return cost, plan, plans
+
+
+def merge_bijections(X, Y, plans, lp_power=2):
```
Comment (rflamary) [inline ot/bsp/bsp_ot.py]:
```suggestion
def merge_bijections(X, Y, perms, p=2):
```

File: .gitmodules [hunk 16]
```diff
@@ -0,0 +1,3 @@
+[submodule "deps/eigen"]
+	path = deps/eigen
+	url = https://gitlab.com/libeigen/eigen.git
```
Comment (rflamary) [inline .gitmodules:3]: please fix the version/commit of eigen here. fdor instance with  branch=5.0

File: RELEASES.md [hunk 17]
```diff
@@ -27,6 +28,7 @@ This new release adds support for sparse cost matrices and a new lazy EMD solver
 - Clean documentation (PR #787)
 - Fix code coverage (PR #791)
 - Fix test of the version of jax in `ot.backend` (PR #794)
+- Reverting the openmp fix on macOS (PR #789) for macOS (PR #796)
```
Comment (rflamary) [inline RELEASES.md]:
```suggestion
```

**Review by rflamary — COMMENTED** [review]
The PR is looking good thanks so much @dcoeurjo and @baptiste-genest . below are  some comments about API.
